packages/logging/src/metadata.js

```diff
@@ -0,0 +1,297 @@
+/*!
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module logging/metadata
+ */
+
+'use strict';
+var is = require('is');
+
+/**
+ * The abstract events emitted to invokers of getDefaultResource and
+ * getProjectId which are fired upon evaluation of the response from the
+ * metadata service.
+ * @const REQUEST_EVENTS
+ */
+var REQUEST_EVENTS = {
+  SUCCESS: 'METADATA_REQUEST_SUCCESS',
+  FAILURE: 'METADATA_REQUEST_FAILURE'
+};
+
+/**
+ * The Metadata class attempts to contact the metadata service and determine,
+ * based on request success and environment variables, what type of resource
+ * the library is operating on.
+ *
+ * @constructor
+ * @alias module:metadata
+ *
+ * @resource [Logs Resource API Documentation]{@link https://cloud.google.com/logging/docs/api/reference/rest/v2/MonitoredResource}
+ *
+ * @param {Duplex} requestStream - A stream returned by a request style
+ *  invocation to the metadata project id endpoint
+ */
+function Metadata(requestStream) {
+  this.metadataProjectId_ = null;
+  this.metadataContactError_ = null;
+  this.metadataRequestStream_ = requestStream;
+  this.attachRequestStreamListeners_(requestStream);
+}
+
+/**
+ * Attaches core request lifecycle event listeners on the request stream
+ * provided to the constuctor.
+ *
+ * @private
+ */
+Metadata.prototype.attachRequestStreamListeners_ = function(stream) {
+  var self = this;
```

Comment:
Any particular reason why? The stream interface decouples the interface nicely, make instrumentation particularly easy, and allows multiple callers to idiomatically listen on the same event without creating extra state inside the class or another 'in-the-middle' event emitter.